GRCBC

anandrdbz opened the pull request:
## Description

Generalized CBC implementation given by Pirozolli and Colonius (2013). Provides the option to supplement  non-reflecting subsonic inflow and subsonic outflow boundary conditions (bc = -7 and -8) with additional terms. Their effect can be turned on by setting bc%grcbc_in and bc%grcbc_out to be true (false by default). Additional option to turn on bc%grcbc_vel_out that includes outflow velocity specification. 

If turned on, inflow boundary conditions require inflow velocities, pressure ,density and volume fraction specifications outside the domain. For outflow conditions, pressure and normal velocity specifications needed, although the normal velocity can be turned off using bc%grcbc_vel_out. 

Refer to case file 2D_ibm_steady_shock for an example with GRCBC inflow conditions for a shock with a given velocity. 

2 additional cases added to the examples directory. 

2D_zero_circ_vortex advects a zero circulation isentropic vortex in the middle of the domain from left to right with Subsonic Inflow and subsonic outflow boundary conditions. 

2D_acoustic_pulse has a radially outward pressure pulse with subsonic outflow boundary conditions on all 4 sides.

Fixes #(issue) [optional]

### Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Something else

### Scope

- [ x] This PR comprises a set of related changes with a common goal

If you cannot check the above box, please split your PR into multiple PRs that each have a common goal.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:

* What computers and compilers did you use to test this:

## Checklist

- [x] I have added comments for the new code
- [x] I added Doxygen docstrings to the new code
- [x] I have made corresponding changes to the documentation (`docs/`)
- [x] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [x] I have added example cases in `examples/` that demonstrate my new feature performing as expected.
They run to completion and demonstrate "interesting physics"
- [x] I ran `./mfc.sh format` before committing my code
- [x] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled
- [x] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [x] I cannot think of a way to condense this code and reduce any introduced additional line count

### If your code changes any code source files (anything in `src/simulation`)

To make sure the code is performing as expected on GPU devices, I have:
- [x] Checked that the code compiles using NVHPC compilers
- [x] Checked that the code compiles using CRAY compilers
- [x] Ran the code on either V100, A100, or H100 GPUs and ensured the new feature performed as expected (the GPU results match the CPU results)
- [x] Ran the code on MI200+ GPUs and ensure the new features performed as expected (the GPU results match the CPU results)
- [ ] Enclosed the new feature via `nvtx` ranges so that they can be identified in profiles
- [ ] Ran a Nsight Systems profile using `./mfc.sh run XXXX --gpu -t simulation --nsys`, and have attached the output file (`.nsys-rep`) and plain text results to this PR
- [ ] Ran an Omniperf profile using `./mfc.sh run XXXX --gpu -t simulation --omniperf`, and have attached the output file and plain text results to this PR.
- [x] Ran my code using various numbers of different GPUs (1, 2, and 8, for example) in parallel and made sure that the results scale similarly to what happens if you run without the new code/feature

sbryngelson commented:
Please fix merge conflicts

anandrdbz commented:
@sbryngelson the checks passed after conflict resolution, so should be ready for merge

sbryngelson commented:
thanks @anandrdbz I will review this soon!

wilfonba commented on examples/2D_acoustic_pulse/case.py:
```diff
@@ -0,0 +1,115 @@
+
+# Configuring case dictionary
+print(json.dumps({
+    # Logistics ================================================================
```
It might be worthwhile to remove the "distracting characters," as mentioned in #609, to save work for whoever fixes it down the road and avoid creating more problems.

sbryngelson replied:
I think these are fine since it will likely be a single find/replace that removes all of these distracting characters from all the case files (and elsewhere.

wilfonba commented on examples/2D_ibm_steady_shock/case.py:
```diff
@@ -0,0 +1,176 @@
+
+# Configuring case dictionary
+print(json.dumps({
+    # Logistics ================================================================
```
It might be worthwhile to remove the "distracting characters," as mentioned in #609, to save work for whoever fixes it down the road and avoid creating more problems.

wilfonba commented on examples/2D_zero_circ_vortex/case.py:
```diff
@@ -0,0 +1,122 @@
+
+# Configuring case dictionary
+print(json.dumps({
+    # Logistics ================================================================
```
It might be worthwhile to remove the "distracting characters," as mentioned in #609, to save work for whoever fixes it down the road and avoid creating more problems.

sbryngelson replied:
I think these are fine since it will likely be a single find/replace that removes all of these distracting characters from all the case files (and elsewhere.

sbryngelson commented on docs/documentation/case.md:
```diff
@@ -805,6 +805,23 @@ The boundary condition supported by the MFC are listed in table [Boundary Condit
+| `bc_[x,y,z]%alpha_rho_in`     | Real Array | Inflow density |
+| `bc_[x,y,z]%alpha_in`         | Real Array | Inflow void fraction |
+
+*: This boundary condition can be used for subsonic inflow (`bc_[x,y,z]%[beg,end]` = -7) and subsonic outflow (`bc_[x,y,z]%[beg,end]` = -8) characteristic boundary conditions. These are based on [Pirozzoli (2013)](references.md#Pirozzoli13). This enables to provide inflow and outflow conditions outside the computational domain. The parameters associated with this feature are listed in table [Gene ...
```
where does the * go?

sbryngelson replied:
it doesn't seem to be attached to any variables

sbryngelson commented on src/simulation/m_cbc.fpp:
```diff
@@ -102,6 +102,18 @@ module m_cbc
+    !$acc declare create(uy_in, uy_out, vy_in, vy_out, wy_in, wy_out, presy_in, presy_out, Dely_in, Dely_out)
+    !$acc declare create(uz_in, uz_out, vz_in, vz_out, wz_in, wz_out, presz_in, presz_out, Delz_in, Delz_out)
+
+    real(kind(0d0)), allocatable, dimension(:) :: alpha_rhox_in, alphax_in
```
same comment as above

sbryngelson commented on src/simulation/m_cbc.fpp:
```diff
@@ -392,6 +404,53 @@ contains
+
+        #:for CBC_DIR, XYZ in [(1, 'x'), (2, 'y'), (3, 'z')]
+            if (${CBC_DIR}$ == 1) then
+                u${XYZ}$_in = bc_${XYZ}$%u_in
```
why use u/v/w when you can use u(1,2,3)?

sbryngelson commented on src/simulation/m_start_up.fpp:
```diff
@@ -1465,6 +1465,9 @@ contains
         !$acc update device(bc_y%vb1, bc_y%vb2, bc_y%vb3, bc_y%ve1, bc_y%ve2, bc_y%ve3)
         !$acc update device(bc_z%vb1, bc_z%vb2, bc_z%vb3, bc_z%ve1, bc_z%ve2, bc_z%ve3)
 
+        !$acc update device(bc_x%grcbc_in, bc_x%grcbc_out, bc_x%grcbc_vel_out)
```
9 variables when you could just have one?

anandrdbz commented:
@sbryngelson I've made most of the suggested changes, the only repetition is in the bc variables, but this is how it's done for all other cases (separate variables for bc_x, bc_y and bc_z). However in the CBC module this is transformed into an array of size (3) as suggested. Most of the verbosity in the code should've been removed now. And I've added 3 more test cases; these should cover every possible line I've added in CBC, so further tests are unnecessary. I've added a few more comments in the CBC file as well

sbryngelson commented:
> @sbryngelson I've made most of the suggested changes, the only repetition is in the bc variables, but this is how it's done for all other cases (separate variables for bc_x, bc_y and bc_z). However in the CBC module this is transformed into an array of size (3) as suggested. Most of the verbosity in the code should've been removed now. And I've added 3 more test cases; these should cover every possible line I've added in CBC, so further tests are unnecessary. I've added a few more comments in the CBC file as well

Thanks I will review again shortly!

anandrdbz commented:
@sbryngelson For some reason the job on Phoenix has been in queue for 2 days, every other test has passed and I've checked the test suite on phoenix on an interactive node

sbryngelson commented:
Yes those nodes are backed up big time for some reason. I'll merge it once it passes you don't have to do anything.
